<a href="https://colab.research.google.com/github/diogo-costa-silva/sql-murder-mystery/blob/main/SQL_Murder_Mystery.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# SQL Murder Mystery

![Sql Murder Mystery image](http://mystery.knightlab.com/174092-clue-illustration.png)

There's been a Murder in SQL City! The SQL Murder Mystery is designed to be both a self-directed lesson to learn SQL concepts and commands and a fun game for experienced SQL users to solve an intriguing crime.

The original statement is as follows: A crime has taken place and the detective needs your help. The detective gave you the crime scene report, but you somehow lost it. You vaguely remember that the crime was a **murder** that occurred sometime on **Jan.15, 2018** and that it took place in **SQL City**.

This mystery can be solved on their [website](http://mystery.knightlab.com/) itself or downloaded from their [github repository](https://github.com/NUKnightLab/sql-mysteries).

**The database's schema**

![Schema](https://raw.githubusercontent.com/NUKnightLab/sql-mysteries/master/schema.png)

In [16]:
# Importing the necessary libraries
import pandas as pd
import sqlite3 as sql

In [17]:
# getting the database from my github
!wget https://github.com/diogo-costa-silva/assets/raw/main/databases/sql-murder-mystery.db -O sql-murder-mystery.db

--2024-02-21 06:52:45--  https://github.com/diogo-costa-silva/assets/raw/main/databases/sql-murder-mystery.db
Resolving github.com (github.com)... 140.82.114.3
Connecting to github.com (github.com)|140.82.114.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/diogo-costa-silva/assets/main/databases/sql-murder-mystery.db [following]
--2024-02-21 06:52:46--  https://raw.githubusercontent.com/diogo-costa-silva/assets/main/databases/sql-murder-mystery.db
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3661824 (3.5M) [application/octet-stream]
Saving to: ‘sql-murder-mystery.db’

sql-murder-mystery. 100%[===================>]   3.49M  --.-KB/s    in 0.07s   

2024-02-21 06:52:46 (49.2 MB/s) - ‘sql-mur

In [18]:
# Setting up a connection to the sqlite database
con = sql.connect('sql-murder-mystery.db')

#seting the dataframe width to max
pd.set_option('display.max_colwidth', None)

## Step 1 - Work with initial statement

Given that from the original statement we have some important data from which to begin our investigation, such as the **type** of crime, the **date** and the place (**city**) where it occurred, we will carry out the first query on the **crime scene report** table.

In [19]:
query_1 = '''
  SELECT *
  FROM crime_scene_report
  WHERE city = "SQL City"
  ORDER BY date;
'''

#running our query
pd.read_sql_query(query_1, con)

date     type  \
0  20170712    theft   
1  20170820    arson   
2  20171110  robbery   
3  20180103  bribery   
4  20180115  assault   
5  20180115  assault   
6  20180115   murder   
7  20180215   murder   
8  20180215   murder   

                                                                                                                                                                                                                                                              description  \
0              A lone hunter stalks the night, firing arrows into the Darkness.\n            There is no hiding, no escape. In the distance, the beast\n            falters, tethered to the void. The killing blow comes without\n            hesitation, without mercy.   
1                                                                                                                                                                  Wield the Hammer of Sol with honor, Titan, it is a thing of\n            legend, both past and future.   
2  The Gjallarhorn shoulder-mounted rocket system was forged from\n            the armor of Guardians who fell at the Twilight Gap. Gifted\n            to the survivors of that terrible battle, the Gjallarhorn\n            is seen as a symbol of honor and survival.   
3                                                                                                                                                 Apparently, Cayde thought it necessary to expose this extremely\n            rare vegetable to a Hive summoning ritual.   
4                                                                                                                                                                                             Hamilton: Lee, do you yield? Burr: You shot him in the side! Yes he yields!   
5                                                                                                                                                                                                                                                        Report Not Found   
6                                                                               Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".   
7                                                                                                                                                                                                                                              REDACTED REDACTED REDACTED   
8                                                                                                                                                                                                                 Someone killed the guard! He took an arrow to the knee!   

       city  
0  SQL City  
1  SQL City  
2  SQL City  
3  SQL City  
4  SQL City  
5  SQL City  
6  SQL City  
7  SQL City  
8  SQL City

From the results from query 1, we can observe that only one record corresponds to the original statement, which mentions the occurrence of a **murder** that occurred sometime on **Jan.15, 2018** and that it took place in **SQL City**. <br>

The description of the occurrence is as follows: <br>

Security footage shows that there were **2 witnesses**. The **first** witness lives at the last house on **"Northwestern Dr"**. The **second** witness, named Annabel, lives somewhere on **"Franklin Ave"**.

Now that we have this additional data, we will continue our investigation.

## Step 2 - Search for the two witnesses of the description found

Let's first search for the first witness, who we already know lives in the last house on "Northwestern Dr.".
In order to determine the last house number in "Northwestern Dr.", you must use the **desc** method in the **ORDER BY** clause.

In [20]:
#checking personal details of both the witnesses
query_2_1 = '''
  SELECT *
  FROM person
  WHERE address_street_name = "Northwestern Dr"
  ORDER BY address_number desc;
'''
pd.read_sql_query(query_2_1, con)[:1]

id            name  license_id  address_number address_street_name  \
0  14887  Morty Schapiro      118009            4919     Northwestern Dr   

         ssn  
0  111564949

As you can see, the discovered record takes us to an individual called "**Morty Schapiro**", where we also have access to his **id**, **license_id** and **ssn**, data that could be useful in the future.



Let's move on to the second witness, who we know is called "Annabel" and lives somewhere on "Franklin Ave".

In [21]:
#checking personal details of both the witnesses
query_2_2 = '''
  SELECT *
  FROM person
  WHERE name like '%Annabel%' AND address_street_name = "Franklin Ave";
'''
pd.read_sql_query(query_2_2, con)[:1]

id            name  license_id  address_number address_street_name  \
0  16371  Annabel Miller      490173             103        Franklin Ave   

         ssn  
0  318771143

Again, through a simple query, we were able to identify the second witness and we now know that her name is "**Annabel Miller**" with **id** 16371 and **license_id** 490173.

## Step 3 - Analyse the interview of both the witnesses taken after the murder

In [22]:
query_3_1 = '''
  SELECT *
  FROM interview
  WHERE person_id = 14887 OR person_id = 16371;
'''
pd.read_sql_query(query_3_1, con)

person_id  \
0      14887   
1      16371   

                                                                                                                                                                                                                        transcript  
0  I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".  
1                                                                                                            I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.

After reviewing the transcripts from both witnesses, we got several clues:

- Killer is a **man** and a **gold member** of the **Get Fit Now Gym** with the membership number "48Z". He left in a car with a plate that <u>included</u> "**H42W**".

- The killer was working out last week on **January 9th**.

## Step 4 - Check the gym database with the new clues from the transcripts

To do this, we will look in the **get_fit_now_check_in** table for records that include **48Z** in the **membership number** and whose **check-in date** is on **January 9th (20180109)**.

In [23]:
query_4_1 = '''
  SELECT *
  FROM get_fit_now_check_in
  WHERE membership_id like "%48Z%" AND check_in_date = 20180109
  ORDER BY check_in_date;
'''
pd.read_sql_query(query_4_1, con)

membership_id  check_in_date  check_in_time  check_out_time
0         48Z7A       20180109           1600            1730
1         48Z55       20180109           1530            1700

As we can see, there are two records that match our search, so we will have to analyze the remaining clues to reduce the number of matching records.

So, let's go to the **drivers_license** table to search for license plate numbers that include **H42W**.

In [24]:
#now, let's check the car details by the above details
query_4_2 = '''
  SELECT *
  FROM drivers_license
  WHERE plate_number like "%H42W%";
'''
pd.read_sql_query(query_4_2, con)

id  age  height eye_color hair_color  gender plate_number   car_make  \
0  183779   21      65      blue     blonde  female       H42W0X     Toyota   
1  423327   30      70     brown      brown    male       0H42W2  Chevrolet   
2  664760   21      71     black      black    male       4H42WR     Nissan   

  car_model  
0     Prius  
1  Spark LS  
2    Altima

This query resulted in 2 records corresponding to men who have **H42W** inserted in their registration number.

We will then proceed to verify the personal details of both individuals.

In [25]:
query_4_3 = '''
  SELECT *
  FROM person
  WHERE license_id = "423327" OR license_id = "664760";
'''
pd.read_sql_query(query_4_3, con)

id            name  license_id  address_number    address_street_name  \
0  51739  Tushar Chandra      664760             312                 Phi St   
1  67318   Jeremy Bowers      423327             530  Washington Pl, Apt 3A   

         ssn  
0  137882671  
1  871539279

We again have 2 records corresponding to the launched query. Which one will be a member of Get Fit Now Gym?

## Step 5 - Check which of this two individuals are a member of the gym


In [26]:
query_5_1 = '''
  SELECT *
  FROM get_fit_now_member
  WHERE person_id = "51739" OR person_id = "67318";
'''
pd.read_sql_query(query_5_1, con)

id  person_id           name  membership_start_date membership_status
0  48Z55      67318  Jeremy Bowers               20160101              gold

After running this last query, we finally have a possible suspect!
So let's check if this name is the culprit of the crime. To do this, in the section of the website that says "Check your solution", we put the name in the corresponding space, as shown in the following figure:

![Check-your-solution](https://raw.githubusercontent.com/diogo-costa-silva/sql-murder-mystery/main/images/check_your_solution_1.png)


But wait! Apparently, there was still another person behind the crime, the real Mastermind. Let's try to find out who this person is to solve this mystery once and for all!

As indicated in the solution, let's start by checking the murderer's transcript.

## Step 6 - Check the murderer's transcript




In [27]:
query_6_1 = '''
  SELECT *
  FROM interview
  WHERE person_id = 67318;
'''
pd.read_sql_query(query_6_1, con)

person_id  \
0      67318   

                                                                                                                                                                                                                                           transcript  
0  I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.\n

The transcript tells us that the real villain is a **woman** with a **Tesla Model S** car that has **red hair**.

Furthermore, she attended the **SQL Symphony Concert** **3** times in **December 2017**.

## Step 7 - Let's follow the physical and car clues first

Let's create a query that follows these clues.

In [28]:
query_7_1 = '''
  SELECT *
  FROM drivers_license
  WHERE car_make = "Tesla" AND car_model = "Model S" AND gender = "female" AND hair_color = "red";
'''
pd.read_sql_query(query_7_1, con)

id  age  height eye_color hair_color  gender plate_number car_make  \
0  202298   68      66     green        red  female       500123    Tesla   
1  291182   65      66      blue        red  female       08CM64    Tesla   
2  918773   48      65     black        red  female       917UU3    Tesla   

  car_model  
0   Model S  
1   Model S  
2   Model S

Taking the **id** from the **drivers_license** table, which we know is the **license_id** from the **person** table, we will consult the latter to check the personal information of these three women:

In [29]:
query_7_2 = '''
  SELECT *
  FROM person
  WHERE license_id = "202298" OR license_id = "291182" OR license_id = "918773";
'''
pd.read_sql_query(query_7_2, con)

id              name  license_id  address_number address_street_name  \
0  78881          Red Korb      918773             107         Camerata Dr   
1  90700     Regina George      291182             332           Maple Ave   
2  99716  Miranda Priestly      202298            1883          Golden Ave   

         ssn  
0  961388910  
1  337169072  
2  987756388

## Step 8 - Let's track attendance at the concert

In [30]:
#checking the event SQL symphony concert
query_8 = '''
  SELECT person_id, count(*), event_name
  FROM facebook_event_checkin
  GROUP BY person_id
  HAVING count(*) = 3 AND event_name = "SQL Symphony Concert" AND date like "%201712%";
'''
pd.read_sql_query(query_8, con)

person_id  count(*)            event_name
0      24556         3  SQL Symphony Concert
1      99716         3  SQL Symphony Concert

Of the only two people who went to the SQL Symphony Concert 3 times in December 2017, one of them has the same **id** as one of the 3 people previously identified with the characteristics described in the murder transcript, **99716**.
This way, we can conclude that this person is most likely the Mastermind we were looking for...

In order to confirm our suspicions, let's check whether or not it is:

![Check-your-solution](https://raw.githubusercontent.com/diogo-costa-silva/sql-murder-mystery/main/images/check_your_solution_2.png)

Confirmed! Our suspect is in fact the Mastermind of this mystery.

We got you Miranda Priestly!!!